File: decentralized accounts/Java/agents/Agent.java
package agents;

import java.util.HashMap;

public class Agent {
	
	//
	// PROPERTIES
	//
	
	HashMap<items.Item, Integer> itemsOwned; 
	
	//
	// CONSTRUCTOR
	//

	public Agent()
	{
		int arraySize = 10;
		
		itemsOwned = new HashMap<items.Item, Integer>(arraySize);
	}
	
	//
	// METHODS
	//
	
	public void addItemType(items.Item item)
	{
		itemsOwned.put(item, 0);
	}
		
	public int getItemQntOwned(items.Item item)
	{
		if(itemsOwned.containsKey(item))
		{
			return itemsOwned.get(item);
		}
		else
		{
			return -1;
		}
	}
	
	public HashMap<items.Item, Integer> getAgentHoldingDistro()
	{
		return itemsOwned;
	}
	
	public void sendItemQnt(agents.Agent receiver, items.Item item, int qnt)
	{
		int currQnt = itemsOwned.get(item);
		itemsOwned.replace(item, currQnt - qnt);
		
		receiver.setItemQnt(item, qnt);
	}
	
	public void setItemQnt(items.Item item, int qnt)
	{
		if(itemsOwned.containsKey(item))
		{
			int currQnt = itemsOwned.get(item);
			itemsOwned.replace(item, currQnt + qnt);
		}
		else
		{
			itemsOwned.put(item, qnt);
		}
	}
	
}

File: decentralized accounts/cpp/src/Agent.cpp
#include "Agent.h"

#include "Item.h"
//#include "MyHashFunction.h"

#include <unordered_map> 

using namespace std;

Agent::Agent(/* args */)
{
}

Agent::~Agent()
{
}

bool Agent::operator==(const Agent& a) const
{
    return &a == this;
}

void Agent::addItemtype(Item item)
{
    //itemsOwned.insert(make_pair(item, 0));
}

int Agent::getItemQntOwned(Item item)
{
    /*
    unordered_map<Item, int>::iterator it = itemsOwned.find(item);
    if (it == itemsOwned.end())
    {
        return -1;
    }
    else
    {
        return it->second;
    }
    */
}

/*
unordered_map<Item, int> Agent::getAgentHoldingDistro()
{
    return itemsOwned;
}
*/

void Agent::sendItemQnt(Agent receiver, Item item, int qnt)
{
    //itemsOwned.at(item) -= qnt;

    //receiver.setItemQnt(item, qnt);
}

void Agent::setItemQnt(Item item, int qnt)
{
    /*
    unordered_map<Item, int>::iterator it = itemsOwned.find(item);
    if (it == itemsOwned.end())
    {
        itemsOwned.insert(make_pair(item, qnt));
    }
    else
    {
        itemsOwned.at(item) += qnt;
    }
    */
}

// HASH FUNCTION
namespace std {
  template <>
  struct hash<Item>
  {
    std::size_t operator()(const Item& k) const
    {
      return (size_t)&k;
    }
  };
}

